wolfcrypt/src/hpke.c

```diff
@@ -0,0 +1,1123 @@
+
+    /* set hpke_suite_id */
+    id = hpke->hpke_suite_id;
+
+    XMEMCPY(id, HPKE_STR, HPKE_STR_LEN);
+    id += HPKE_STR_LEN;
+
+    if (ret == 0) {
+      ret = I2OSP(kem, 2, id);
+      id += 2;
+    }
+    if (ret == 0) {
+      ret = I2OSP(kdf, 2, id);
+      id += 2;
+    }
+    if (ret == 0) {
+      ret = I2OSP(aead, 2, id);
+    }
+    if (ret != 0)
+        return ret;
+
+    switch (kem)
+    {
+        case DHKEM_P256_HKDF_SHA256:
+            hpke->curve_id = ECC_SECP256R1;
+            hpke->Nsecret = WC_SHA256_DIGEST_SIZE;
+            hpke->Nh = WC_SHA256_DIGEST_SIZE;
+            hpke->Ndh = wc_ecc_get_curve_size_from_id( hpke->curve_id );
+            hpke->Npk = 1 + hpke->Ndh * 2;
+            break;
+
+        case DHKEM_P384_HKDF_SHA384:
+            hpke->curve_id = ECC_SECP384R1;
+            hpke->Nsecret = WC_SHA384_DIGEST_SIZE;
+            hpke->Nh = WC_SHA384_DIGEST_SIZE;
+            hpke->Ndh = wc_ecc_get_curve_size_from_id( hpke->curve_id );
+            hpke->Npk = 1 + hpke->Ndh * 2;
+            break;
+
+        case DHKEM_P521_HKDF_SHA512:
+            hpke->curve_id = ECC_SECP521R1;
+            hpke->Nsecret = WC_SHA512_DIGEST_SIZE;
+            hpke->Nh = WC_SHA512_DIGEST_SIZE;
+            hpke->Ndh = wc_ecc_get_curve_size_from_id( hpke->curve_id );
+            hpke->Npk = 1 + hpke->Ndh * 2;
+            break;
+
+        case DHKEM_X25519_HKDF_SHA256:
+            hpke->Nsecret = WC_SHA256_DIGEST_SIZE;
+            hpke->Nh = WC_SHA256_DIGEST_SIZE;
+            hpke->Ndh = 32;
+            hpke->Npk = 32;
+            /* hpke->curve_id = ECC_X25519; */
+            break;
+
+        /* TODO: Add X448 */
+        case DHKEM_X448_HKDF_SHA512:
+            hpke->Nsecret = WC_SHA512_DIGEST_SIZE;
+            hpke->Nh = WC_SHA512_DIGEST_SIZE;
+            hpke->Ndh = 64;
+            hpke->Npk = 56;
+            /* hpke->curve_id = ECC_X448; */
+            break;
+
+        default:
+            ret = BAD_FUNC_ARG;
+            break;
+    }
+
+    switch (kdf)
+    {
+        case HKDF_SHA256:
+            hpke->kdf_digest = WC_SHA256;
+            break;
+
+        case HKDF_SHA384:
+            hpke->kdf_digest = WC_SHA384;
+            break;
+
+        case HKDF_SHA512:
+            hpke->kdf_digest = WC_SHA512;
+            break;
+
+        default:
+            ret = BAD_FUNC_ARG;
+            break;
+    }
+
+    switch (aead)
+    {
+        case HPKE_AES_128_GCM:
+            hpke->Nk = AES_128_KEY_SIZE;
+            hpke->Nn = GCM_NONCE_MID_SZ;
+            hpke->Nt = AES_BLOCK_SIZE;
+            break;
+
+        case HPKE_AES_256_GCM:
+            hpke->Nk = AES_256_KEY_SIZE;
+            hpke->Nn = GCM_NONCE_MID_SZ;
+            hpke->Nt = AES_BLOCK_SIZE;
+            break;
+
+        default:
+            ret = BAD_FUNC_ARG;
+            break;
+    }
+
+    return 0;
+}
+
+int wc_HpkeGenerateKeyPair(Hpke* hpke, void** keypair)
+{
+    int ret;
+#ifdef WOLFSSL_SMALL_STACK
```

Comment:
Should be passed in if possible.
Initialising an RNG is really expensive, especially int FIPS.